hbase-client/src/main/java/org/apache/hadoop/hbase/zookeeper/ReadOnlyZKClient.java

```diff
@@ -258,6 +266,22 @@ public void closed(IOException e) {
     }
   }
 
+  public CompletableFuture<byte[]> getWithTimeout(String path, long endTime, HashedWheelTimer retryTimer) {
+    CompletableFuture<byte[]> future = get(path);
+    TimerTask timerTask = new TimerTask() {
+      @Override
+      public void run(Timeout timeout) throws Exception {
+        if (!future.isCancelled() && !future.isDone() && !future.isCompletedExceptionally()) {
+          future.completeExceptionally(
+            new DoNotRetryIOException("Zookeeper get could not be completed by " + endTime));
+        }
+      }
+    };
+
+    retryTimer.newTimeout(timerTask, endTime + 1, TimeUnit.MILLISECONDS);
+    return future;
+  }
+
```

Comment:
We have some opportunity to refactor common TimerTask creation, let's do something like this:

```
  public CompletableFuture<byte[]> getWithTimeout(final String path, final long timeoutMs,
    final HashedWheelTimer retryTimer) {
    CompletableFuture<byte[]> future = get(path);
    TimerTask timerTask = getTimerTask(timeoutMs, future, "GET");
    retryTimer.newTimeout(timerTask, timeoutMs + 1, TimeUnit.MILLISECONDS);
    return future;
  }

  public CompletableFuture<Stat> existsWithTimeout(final String path, final long timeoutMs,
    final HashedWheelTimer retryTimer) {
    CompletableFuture<Stat> future = exists(path);
    TimerTask timerTask = getTimerTask(timeoutMs, future, "EXISTS");
    retryTimer.newTimeout(timerTask, timeoutMs + 1, TimeUnit.MILLISECONDS);
    return future;
  }

  public CompletableFuture<List<String>> listWithTimeout(final String path, final long timeoutMs,
    final HashedWheelTimer retryTimer) {
    CompletableFuture<List<String>> future = list(path);
    TimerTask timerTask = getTimerTask(timeoutMs, future, "LIST");
    retryTimer.newTimeout(timerTask, timeoutMs + 1, TimeUnit.MILLISECONDS);
    return future;
  }

  private static TimerTask getTimerTask(final long timeoutMs, final CompletableFuture<?> future,
    final String api) {
    return timeout -> {
      if (!future.isCancelled() && !future.isDone() && !future.isCompletedExceptionally()) {
        future.completeExceptionally(new DoNotRetryIOException(
          "Zookeeper " + api + " could not be completed in " + timeoutMs + " ms"));
      }
    };
  }

```